issue with grant on fields

see: 

- https://github.com/voedger/voedger/blob/3499d71fa8f96f78a51b3a867cc90cb51905e8b4/pkg/parser/impl_analyse.go#L376 
- https://github.com/voedger/voedger/blob/3499d71fa8f96f78a51b3a867cc90cb51905e8b4/pkg/parser/impl_analyse.go#L412

The following test fails:
```go
      t.Run("Issue 3476", func(t *testing.T) {
		schema, err := require.AppSchema(`APPLICATION test();
			WORKSPACE W (
				ROLE role;
				TABLE t INHERITS sys.CDoc(
					number int32,
					name varchar
				);
				GRANT 
					SELECT(sys.ID, number, name),
					UPDATE(number, name)
				ON TABLE t TO role;
			);
		`)
		require.NoError(err)
		builder := builder.New()
		require.NoError(BuildAppDefs(schema, builder))
		app, err := builder.Build()
		require.NoError(err)

		t.Log(app.ACL())
		require.Len(app.ACL(), 2)
	})
```

Output:
```
--- FAIL: Test_Grants_Inherit (0.01s)
    /Users/michael/Workspaces/michael-say/voedger/pkg/parser/impl_test.go:2439: 
        	Error Trace:	/Users/michael/Workspaces/michael-say/voedger/pkg/parser/impl_test.go:2439
        	Error:      	"[GRANT [Update Select] ON QNAMES(pkg.t)[sys.ID number name number name] TO pkg.role]" should have 2 item(s), but has 1
        	Test:       	Test_Grants_Inherit
    --- FAIL: Test_Grants_Inherit/Issue_3476 (0.01s)
        /Users/michael/Workspaces/michael-say/voedger/pkg/parser/impl_test.go:2438: [GRANT [Update Select] ON QNAMES(pkg.t)[sys.ID number name number name] TO pkg.role]
```

Note that it returns 1 ACL rule with both "Select" and "Update" operations and the strange set of fields.

--- a/pkg/parser/impl_analyse.go
+++ b/pkg/parser/impl_analyse.go
@@ -223,11 +223,11 @@ func analyseGrantOrRevoke(toOrFrom DefQName, grant *GrantOrRevoke, c *iterateCtx
 			checkColumn := func(column Identifier) error {
 				for _, f := range view.Items {
 					if f.Field != nil && f.Field.Name.Value == column.Value {
-						grant.columns = append(grant.columns, string(column.Value))
+						grant.View.columns = append(grant.View.columns, string(column.Value))
 						return nil
 					}
 					if f.RefField != nil && f.RefField.Name.Value == column.Value {
-						grant.columns = append(grant.columns, string(column.Value))
+						grant.View.columns = append(grant.View.columns, string(column.Value))
 						return nil
 					}
 				}
@@ -373,27 +373,14 @@ func analyseGrantOrRevoke(toOrFrom DefQName, grant *GrantOrRevoke, c *iterateCtx
 			items = descriptor.Items
 		}
 		grant.Table.Table.qName = pkg.NewQName(Ident(named.GetName()))
-		for _, item := range grant.Table.Items {
-			if item.Insert {
-				grant.ops = append(grant.ops, appdef.OperationKind_Insert)
-			} else if item.Update {
-				grant.ops = append(grant.ops, appdef.OperationKind_Update)
-			} else if item.Select {
-				grant.ops = append(grant.ops, appdef.OperationKind_Select)
-			} else if item.Activate {
-				grant.ops = append(grant.ops, appdef.OperationKind_Activate)
-			} else if item.Deactivate {
-				grant.ops = append(grant.ops, appdef.OperationKind_Deactivate)
-			}
-		}
+
+		// Helper function to check if a column exists and add it to the appropriate operation
 		checkColumn := func(column Ident) error {
 			for _, f := range items {
 				if f.Field != nil && f.Field.Name == column {
-					grant.columns = append(grant.columns, string(column))
 					return nil
 				}
 				if f.RefField != nil && f.RefField.Name == column {
-					grant.columns = append(grant.columns, string(column))
 					return nil
 				}
 				if f.NestedTable != nil && f.NestedTable.Name == column {
@@ -402,23 +389,58 @@ func analyseGrantOrRevoke(toOrFrom DefQName, grant *GrantOrRevoke, c *iterateCtx
 			}
 			return ErrUndefinedField(string(column))
 		}
+
+		// Handle ALL case
 		if grant.Table.All != nil {
+			// For each operation, process the columns
 			for _, column := range grant.Table.All.Columns {
 				if err := checkColumn(column.Value); err != nil {
 					c.stmtErr(&column.Pos, err)
+					return
 				}
+				grant.Table.All.columns = append(grant.Table.All.columns, string(column.Value))
 			}
-		}
-		for _, i := range grant.Table.Items {
-			for _, column := range i.Columns {
-				if column.Name != nil {
-					if err := checkColumn(column.Name.Value); err != nil {
-						c.stmtErr(&column.Pos, err)
-					}
+		} else if len(grant.Table.Items) > 0 {
+			opColumns := make(map[appdef.OperationKind][]string)
+			// Handle operation-specific grants
+			for _, item := range grant.Table.Items {
+				var op appdef.OperationKind
+				if item.Insert {
+					op = appdef.OperationKind_Insert
+				} else if item.Update {
+					op = appdef.OperationKind_Update
+				} else if item.Select {
+					op = appdef.OperationKind_Select
+				} else if item.Activate {
+					op = appdef.OperationKind_Activate
+				} else if item.Deactivate {
+					op = appdef.OperationKind_Deactivate
 				} else {
-					grant.columns = append(grant.columns, column.SysName)
+					continue
+				}
+
+				// Process columns for this specific operation
+				if len(item.Columns) == 0 {
+					opColumns[op] = []appdef.FieldName{}
+				} else {
+					for _, column := range item.Columns {
+						if column.Name != nil {
+							if err := checkColumn(column.Name.Value); err != nil {
+								c.stmtErr(&column.Pos, err)
+								return
+							}
+							opColumns[op] = append(opColumns[op], string(column.Name.Value))
+						} else {
+							// System column - add to operation's list
+							opColumns[op] = append(opColumns[op], column.SysName)
+							// // Also maintain backward compatibility
+							// grant.columns = append(grant.columns, column.SysName)
+						}
+					}
 				}
 			}
+			// Store the operation-to-columns mapping in the grant
+			grant.opColumns = opColumns
 		}
 
 	}
@@ -904,7 +926,7 @@ func checkStorageEntity(key *StateStorage, f *StorageStmt, c *iterateCtx) error 
 		}
 		for _, entity := range key.Entities {
 			if err2 := resolveInCtx(entity, c, func(view *ViewStmt, pkg *PackageSchemaAST) error {
-				key.entityQNames = append(key.entityQNames, pkg.NewQName(entity.Name))
+				key.entityQNames = append(key.entityQNames, pkg.NewQName(view.Name))
 				return nil
 			}); err2 != nil {
 				return err2

--- a/pkg/parser/impl_build.go
+++ b/pkg/parser/impl_build.go
@@ -353,7 +353,7 @@ func (c *buildContext) grantsAndRevokes() error {
 				}
 
 				if s.Revoke.View != nil {
-					wsb.Grant(s.Revoke.ops, s.Revoke.filter(), s.Revoke.View.columns, s.Revoke.toRole, comments...)
+					wsb.Revoke(s.Revoke.ops, s.Revoke.filter(), s.Revoke.View.columns, s.Revoke.toRole, comments...)
 					continue
 				}
 

--- a/pkg/parser/impl_build.go
+++ b/pkg/parser/impl_build.go
@@ -304,11 +304,7 @@ func (c *buildContext) grantsAndRevokes() error {
 
 				if s.Grant.Table != nil && s.Grant.Table.Items != nil {
 					for op, columns := range s.Grant.opColumns {
-						fieldNames := make([]appdef.FieldName, len(columns))
-						for i, col := range columns {
-							fieldNames[i] = col
-						}
-						wsb.Grant([]appdef.OperationKind{op}, s.Grant.filter(), fieldNames, s.Grant.toRole, comments...)
+						wsb.Grant([]appdef.OperationKind{op}, s.Grant.filter(), columns, s.Grant.toRole, comments...)
 					}
 					continue
 				}
@@ -343,11 +339,7 @@ func (c *buildContext) grantsAndRevokes() error {
 
 				if s.Revoke.Table != nil && s.Revoke.Table.Items != nil {
 					for op, columns := range s.Revoke.opColumns {
-						fieldNames := make([]appdef.FieldName, len(columns))
-						for i, col := range columns {
-							fieldNames[i] = col
-						}
-						wsb.Revoke([]appdef.OperationKind{op}, s.Revoke.filter(), fieldNames, s.Revoke.toRole, comments...)
+						wsb.Revoke([]appdef.OperationKind{op}, s.Revoke.filter(), columns, s.Revoke.toRole, comments...)
 					}
 					continue
 				}

--- a/pkg/parser/impl_build.go
+++ b/pkg/parser/impl_build.go
@@ -289,31 +289,79 @@ func (c *buildContext) grantsAndRevokes() error {
 			if s.Grant != nil {
 				wsb := s.Grant.workspace.mustBuilder(c)
 				comments := s.Grant.GetComments()
+
+				// Handle ALL cases
 				if (s.Grant.AllTablesWithTag != nil && s.Grant.AllTablesWithTag.All) ||
-					(s.Grant.Table != nil && s.Grant.Table.All != nil) ||
 					(s.Grant.AllTables != nil && s.Grant.AllTables.All) {
 					wsb.Grant(grantAllToTableOps, s.Grant.filter(), []appdef.FieldName{}, s.Grant.toRole, comments...)
 					continue
 				}
-				wsb.Grant(s.Grant.ops, s.Grant.filter(), s.Grant.columns, s.Grant.toRole, comments...)
+
+				if s.Grant.Table != nil && s.Grant.Table.All != nil {
+					wsb.Grant(grantAllToTableOps, s.Grant.filter(), s.Grant.Table.All.columns, s.Grant.toRole, comments...)
+					continue
+				}
+
+				if s.Grant.Table != nil && s.Grant.Table.Items != nil {
+					for op, columns := range s.Grant.opColumns {
+						fieldNames := make([]appdef.FieldName, len(columns))
+						for i, col := range columns {
+							fieldNames[i] = appdef.FieldName(col)
+						}
+						wsb.Grant([]appdef.OperationKind{op}, s.Grant.filter(), fieldNames, s.Grant.toRole, comments...)
+					}
+					continue
+				}
+
+				if s.Grant.View != nil {
+					wsb.Grant(s.Grant.ops, s.Grant.filter(), s.Grant.View.columns, s.Grant.toRole, comments...)
+					continue
+				}
+
+				wsb.Grant(s.Grant.ops, s.Grant.filter(), []appdef.FieldName{}, s.Grant.toRole, comments...)
 			}
 		}
 	}
+
 	revokes := func(stmts []WorkspaceStatement) {
 		for _, s := range stmts {
 			if s.Revoke != nil {
 				wsb := s.Revoke.workspace.mustBuilder(c)
 				comments := s.Revoke.GetComments()
+
+				// Handle ALL cases
 				if (s.Revoke.AllTablesWithTag != nil && s.Revoke.AllTablesWithTag.All) ||
-					(s.Revoke.Table != nil && s.Revoke.Table.All != nil) ||
 					(s.Revoke.AllTables != nil && s.Revoke.AllTables.All) {
 					wsb.Revoke(grantAllToTableOps, s.Revoke.filter(), []appdef.FieldName{}, s.Revoke.toRole, comments...)
 					continue
 				}
-				wsb.Revoke(s.Revoke.ops, s.Revoke.filter(), s.Revoke.columns, s.Revoke.toRole, comments...)
+
+				if s.Revoke.Table != nil && s.Revoke.Table.All != nil {
+					wsb.Revoke(grantAllToTableOps, s.Revoke.filter(), s.Revoke.Table.All.columns, s.Revoke.toRole, comments...)
+					continue
+				}
+
+				if s.Revoke.Table != nil && s.Revoke.Table.Items != nil {
+					for op, columns := range s.Revoke.opColumns {
+						fieldNames := make([]appdef.FieldName, len(columns))
+						for i, col := range columns {
+							fieldNames[i] = appdef.FieldName(col)
+						}
+						wsb.Revoke([]appdef.OperationKind{op}, s.Revoke.filter(), fieldNames, s.Revoke.toRole, comments...)
+					}
+					continue
+				}
+
+				if s.Revoke.View != nil {
+					wsb.Grant(s.Revoke.ops, s.Revoke.filter(), s.Revoke.View.columns, s.Revoke.toRole, comments...)
+					continue
+				}
+
+				wsb.Revoke(s.Revoke.ops, s.Revoke.filter(), []appdef.FieldName{}, s.Revoke.toRole, comments...)
 			}
 		}
 	}
+
 	handleWorkspace := func(stmts []WorkspaceStatement) {
 		grants(stmts)
 		revokes(stmts)

--- a/pkg/parser/impl_build.go
+++ b/pkg/parser/impl_build.go
@@ -306,7 +306,7 @@ func (c *buildContext) grantsAndRevokes() error {
 					for op, columns := range s.Grant.opColumns {
 						fieldNames := make([]appdef.FieldName, len(columns))
 						for i, col := range columns {
-							fieldNames[i] = appdef.FieldName(col)
+							fieldNames[i] = col
 						}
 						wsb.Grant([]appdef.OperationKind{op}, s.Grant.filter(), fieldNames, s.Grant.toRole, comments...)
 					}
@@ -345,7 +345,7 @@ func (c *buildContext) grantsAndRevokes() error {
 					for op, columns := range s.Revoke.opColumns {
 						fieldNames := make([]appdef.FieldName, len(columns))
 						for i, col := range columns {
-							fieldNames[i] = appdef.FieldName(col)
+							fieldNames[i] = col
 						}
 						wsb.Revoke([]appdef.OperationKind{op}, s.Revoke.filter(), fieldNames, s.Revoke.toRole, comments...)
 					}

--- a/pkg/parser/impl_test.go
+++ b/pkg/parser/impl_test.go
@@ -2373,6 +2373,67 @@ func Test_Grants(t *testing.T) {
 		require.Equal(1, numACLs)
 	})
 
+	t.Run("Grant ops with different fields", func(t *testing.T) {
+		schema, err := require.AppSchema(`APPLICATION test();
+			WORKSPACE W (
+				ROLE role;
+				TABLE t INHERITS sys.CDoc(
+					number int32,
+					name varchar
+				);
+				GRANT 
+					SELECT(sys.ID, number, name),
+					UPDATE(number, name)
+				ON TABLE t TO role;
+			);
+		`)
+		require.NoError(err)
+		builder := builder.New()
+		require.NoError(BuildAppDefs(schema, builder))
+		app, err := builder.Build()
+		require.NoError(err)
+
+		t.Log(app.ACL())
+		require.Len(app.ACL(), 2)
+		for _, acl := range app.ACL() {
+			require.Equal(appdef.PolicyKind_Allow, acl.Policy())
+			require.Equal("pkg.role", acl.Principal().QName().String())
+			if acl.Ops()[0] == appdef.OperationKind_Select {
+				require.EqualValues([]appdef.OperationKind{appdef.OperationKind_Select}, acl.Ops())
+				require.EqualValues([]string{"sys.ID", "number", "name"}, acl.Filter().Fields())
+			} else {
+				require.EqualValues([]appdef.OperationKind{appdef.OperationKind_Update}, acl.Ops())
+				require.EqualValues([]string{"number", "name"}, acl.Filter().Fields())
+			}
+		}
+	})
+
+	t.Run("Grant all with fields", func(t *testing.T) {
+		schema, err := require.AppSchema(`APPLICATION test();
+			WORKSPACE W (
+				ROLE role;
+				TABLE t INHERITS sys.CDoc(
+					number int32,
+					name varchar
+				);
+				GRANT ALL(number, name) ON TABLE t TO role;
+			);
+		`)
+		require.NoError(err)
+		builder := builder.New()
+		require.NoError(BuildAppDefs(schema, builder))
+		app, err := builder.Build()
+		require.NoError(err)
+
+		t.Log(app.ACL())
+		require.Len(app.ACL(), 1)
+		for _, acl := range app.ACL() {
+			require.Equal(appdef.PolicyKind_Allow, acl.Policy())
+			require.Equal("pkg.role", acl.Principal().QName().String())
+			require.EqualValues([]appdef.OperationKind{appdef.OperationKind_Insert, appdef.OperationKind_Update, appdef.OperationKind_Select}, acl.Ops())
+			require.EqualValues([]string{"number", "name"}, acl.Filter().Fields())
+		}
+	})
 }
 
 func Test_Grants_Inherit(t *testing.T) {

--- a/pkg/parser/types.go
+++ b/pkg/parser/types.go
@@ -730,7 +730,7 @@ type GrantOrRevoke struct {
 	/* filled on the analysis stage */
 	toRole    appdef.QName
 	ops       []appdef.OperationKind
-	opColumns map[appdef.OperationKind][]string
+	opColumns map[appdef.OperationKind][]appdef.FieldName
 
 	workspace workspaceAddr
 }

--- a/pkg/parser/types.go
+++ b/pkg/parser/types.go
@@ -676,8 +676,9 @@ type GrantAllTablesAction struct {
 
 type GrantTableAll struct {
 	Pos      lexer.Position
-	GrantAll bool         `parser:"@'ALL'"`
-	Columns  []Identifier `parser:"( '(' @@ (',' @@)* ')' )?"`
+	GrantAll bool               `parser:"@'ALL'"`
+	Columns  []Identifier       `parser:"( '(' @@ (',' @@)* ')' )?"`
+	columns  []appdef.FieldName // filled on the analysis stage
 }
 
 type GrantTableActions struct {
@@ -703,9 +704,10 @@ type GrantAllTables struct {
 
 type GrantView struct {
 	Pos        lexer.Position
-	AllColumns bool         `parser:"(@(SELECT ONVIEW) | "`
-	Columns    []Identifier `parser:"( SELECT '(' @@ (',' @@)* ')' ONVIEW))"`
-	View       DefQName     `parser:"@@"`
+	AllColumns bool               `parser:"(@(SELECT ONVIEW) | "`
+	Columns    []Identifier       `parser:"( SELECT '(' @@ (',' @@)* ')' ONVIEW))"`
+	View       DefQName           `parser:"@@"`
+	columns    []appdef.FieldName // filled on the analysis stage
 }
 
 type GrantOrRevoke struct {
@@ -728,7 +730,8 @@ type GrantOrRevoke struct {
 	/* filled on the analysis stage */
 	toRole    appdef.QName
 	ops       []appdef.OperationKind
-	columns   []appdef.FieldName
+	opColumns map[appdef.OperationKind][]string
+
 	workspace workspaceAddr
 }
 

